### ISSUE TITLE
SQL lexer doesn't understand dollar-quoted values

### ISSUE THREAD

**Issue by rkneufeld** [weavejester/ragtime#38]
Hey there. I'm working on a postgres-based application that makes use of stored procedures, and I have encountered what I believe is a limitation of ragtime's lexer. In particular, the lexer does not seem to be able to properly split on dollar-quoted bodies.

Take for example this migration file:

```sql
DROP FUNCTION IF EXISTS add_thing_to_partition(int, int);

CREATE OR REPLACE FUNCTION add_thing_to_partition(pid int, tid int)
 RETURNS VOID AS
$$
DECLARE
  existing_partition   RECORD;
BEGIN

    SELECT * into existing_partition FROM partition_thingss WHERE partition_id = pid AND thing_id = tid;

    IF not found THEN
       INSERT INTO partition_things values(pid, tid);
    END IF;

END;
$$
LANGUAGE plpgsql;
```

As it currently stands, [mark-sql-statement-ends](https://github.com/weavejester/ragtime/blob/master/ragtime.sql.files/src/ragtime/sql/files.clj#L51-L58) ignores `$$` and splits on `;`, causing the following broken SQL statements:

* `"DROP FUNCTION IF EXISTS add_thing_to_partition(int, int)"`
* `"CREATE OR REPLACE FUNCTION add_thing_to_partition(pid int, bid int)\n RETURNS VOID AS\n$$ \\\nDECLARE \\\n  existing_partition   RECORD"`
* `"\\\nBEGIN \\\n \\\n    SELECT * into existing_partition FROM partition_things WHERE partition_id = pid AND thing_id = bid"`
* `"\\\n \\\n    IF not found THEN \\\n       INSERT INTO partition_things values(pid, bid)"`
* `"\\\n    END IF"`
* `"\\\n \\\nEND"`
* `"\\\n$$\nLANGUAGE plpgsql"`

---

Replacing `$$` with `'` corrects the problem, but this is not ideal. Single-quotes lack some of the capabilities for escaping and nesting that dollar-quotes allow for. For example, dollar-quotes can be named and nested ([docs](http://www.postgresql.org/docs/9.3/interactive/sql-syntax-lexical.html)).

**Comment by ckuttruff** [comment]
@rkneufeld I had similar issues with the parsing of stored procedures... it motivated me to take a different approach and avoid the complexity of parsing sql files... you might take a look at this ( https://github.com/ckuttruff/clj-sql-up ) - which works as you'd expect with more complicated sql syntax

**Comment by weavejester** [comment]
Out of interest, @ckuttruff, what made you decide to create a migration plugin rather than hook into Ragtime?

**Comment by ckuttruff** [comment]
Hey @weavejester 
I originally had that intention with this PR ( https://github.com/weavejester/ragtime/pull/18 ), but it seemed like not the best approach given the way things were currently structured with ragtime and the expectation of specifying migrations as sql files.  The way of modeling migration statements as a vector of sql strings seemed much simpler (avoiding the necessary parsing logic), but obviously would be a breaking paradigm shift for ragtime.

Apologies if this was perceived as a lack of desire to contribute to ragtime... that is certainly not the case.  Given my limited clojure experience, and the need to maintain backwards compatibility, it was much simpler for me to attack the issue from a blank slate

**Comment by weavejester** [comment]
Ragtime doesn't have an opinion on how migrations are specified. You just need to provide it with a function that returns a sequence of migration maps.

Presumably you wrote clj-sql-up from scratch because you were unaware of that?

**Comment by ckuttruff** [comment]
hmm, well I actually wrote clj-sql-up for a variety of reasons; it was a great learning opportunity amongst other things.

But no, I was not aware aware of an alternative method for specifying migration content beyond sql files; was just going off the documentation here: https://github.com/weavejester/ragtime/wiki/Getting-Started#migrations20120903221323-add-test-tableupsql which just specifies creating plain sql files under a migrations directory.

Thanks for the prompt response and sorry if there was more I could have done to contribute to ragtime; just wanted to mention an alternate solution on this issue thread that doesn't run into some of the parsing issues that @rkneufeld and others have encountered in this particular situation.

**Comment by weavejester** [comment]
No need to apologise! I was just curious as to your reasoning, as that would give me an idea of what could be improved.

The Ragtime README could do a better job at explaining the library, and how different migration strategies could be used.

**Comment by weavejester** [comment]
Should be fixed by version 0.4.0 redesign.
